### PR TITLE
Fetch filters from remote if local cache is invalid

### PR DESCRIPTION
**Pull request by ischasny**
On startup immediately refresh local caches from remote if they are invalid. That might happen due to a previously broken connection that has prevented the process from starting up.

Fixes https://github.com/filecoin-project/boost/issues/1142

### CODE REVIEW

**Comment by dirkmc** [comment]
Thanks a lot for diving in to fix this bug!

I wonder if it would make sense to simply not write to the cache file until after it has successfully been parsed? Then we don't need to worry about adding extra logic in the `Start()` method.

**Comment by ischasny** [comment]
> Thanks a lot for diving in to fix this bug!
> 
> I wonder if it would make sense to simply not write to the cache file until after it has successfully been parsed? Then we don't need to worry about adding extra logic in the `Start()` method.

As I understand it was a deliberate implementation choice to write [alongside](https://github.com/filecoin-project/boost/blob/main/cmd/booster-bitswap/filters/filters.go#L102) reading as the denylist might be quite large (let me know if I misunderstood the design intent). I can change to write in bulk but then there will be memory implication.

**Comment by dirkmc** [comment]
Good point - I wonder if we could write to a temp file, then rename it to the cache file on success?

**Comment by ischasny** [comment]
Good point - will add that soon. Would you not want to still verify cache correctness on start though? So that the process can self recover for example if data format is going to change in the future?

**Comment by ischasny** [comment]
> I think I'd rather keep the caching logic in one place, and then if we change formats in future we can add a migration stage at startup

Cool - pushed the updates in

File: cmd/booster-bitswap/filters/filters.go [hunk 1]
```diff
@@ -93,12 +93,21 @@ func (f *filter) update() error {
...
-	cache, err := os.OpenFile(f.CacheFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	// write cache to the temp file first
+	tmpFile := f.CacheFile + ".tmp"
+	cache, err := os.OpenFile(tmpFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
 	if err != nil {
 		return fmt.Errorf("opening cache file: %w", err)
 	}
-	defer cache.Close()
+	defer func() {
+		cache.Close()
```
Comment (dirkmc) [inline cmd/booster-bitswap/filters/filters.go:103]: Should we check the return value of cache.Close() also?

**Review by dirkmc — COMMENTED** [review]
Looks great!
Just one question then I think we're ready for merge.

File: cmd/booster-bitswap/filters/filters.go (hunk 1 above)
Comment (ischasny) [reply cmd/booster-bitswap/filters/filters.go]: Great spot, fixed.

**Comment by dirkmc** [comment]
hmm looks like `TestMultiFilterCacheShouldNotBeSavedIfDataWasInvalid` is now failing 🤔

**Comment by ischasny** [comment]
> hmm looks like `TestMultiFilterCacheShouldNotBeSavedIfDataWasInvalid` is now failing 🤔

Yes, on it - small mistake

**Review by dirkmc — APPROVED** [review]
Looks great, thanks Ivan! ✅

**Pull request merged**